heron/scheduler-core/src/java/com/twitter/heron/scheduler/UpdateTopologyManager.java

```diff
@@ -0,0 +1,249 @@
+   * @param proposedProtoPackingPlan packing plan to change the topology to
+   */
+  public void updateTopology(final PackingPlans.PackingPlan existingProtoPackingPlan,
+                             final PackingPlans.PackingPlan proposedProtoPackingPlan)
+      throws ExecutionException, InterruptedException {
+    String topologyName = Runtime.topologyName(runtime);
+    PackingPlan existingPackingPlan = deserializer.fromProto(existingProtoPackingPlan);
+    PackingPlan proposedPackingPlan = deserializer.fromProto(proposedProtoPackingPlan);
+
+    assertTrue(proposedPackingPlan.getContainers().size() > 0,
+        "proposed packing plan must have at least 1 container %s", proposedPackingPlan);
+
+    ContainerDelta containerDelta = new ContainerDelta(
+        existingPackingPlan.getContainers(), proposedPackingPlan.getContainers());
+    int newContainerCount = containerDelta.getContainersToAdd().size();
+    int removableContainerCount = containerDelta.getContainersToRemove().size();
+
+    String message = String.format("Topology change requires %s new containers and removing %s "
+            + "existing containers, but the scheduler does not support scaling, aborting. "
+            + "Existing packing plan: %s, proposed packing plan: %s",
+        newContainerCount, removableContainerCount, existingPackingPlan, proposedPackingPlan);
+    assertTrue(newContainerCount + removableContainerCount == 0 || scalableScheduler.isPresent(),
+        message);
+
+    SchedulerStateManagerAdaptor stateManager = Runtime.schedulerStateManagerAdaptor(runtime);
+
+    // assert found PackingPlan is same as existing PackingPlan.
+    validateCurrentPackingPlan(existingProtoPackingPlan, topologyName, stateManager);
+
+    // fetch the topology, which will need to be updated
+    TopologyAPI.Topology updatedTopology =
+        getUpdatedTopology(topologyName, proposedPackingPlan, stateManager);
+
+    // request new resources if necessary. Once containers are allocated we should make the changes
+    // to state manager quickly, otherwise the scheduler might penalize for thrashing on start-up
+    if (newContainerCount > 0) {
+      scalableScheduler.get().addContainers(containerDelta.getContainersToAdd());
+    }
+
+    // update parallelism in updatedTopology since TMaster checks that
+    // Sum(parallelism) == Sum(instances)
+    logFine("Deleted existing Topology: %s", stateManager.deleteTopology(topologyName));
+    logFine("Set new Topology: %s", stateManager.setTopology(updatedTopology, topologyName));
+
+    // update packing plan to trigger the scaling event
+    logFine("Deleted existing packing plan: %s", stateManager.deletePackingPlan(topologyName));
+    logFine("Set new PackingPlan: %s",
+        stateManager.setPackingPlan(proposedProtoPackingPlan, topologyName));
+
+    // delete the physical plan so TMaster doesn't try to re-establish it on start-up.
+    logFine("Deleted Physical Plan: %s", stateManager.deletePhysicalPlan(topologyName));
```

Comment:
In the past we discussed about issues when multiple components were updating a `PackingPlan` (`Launcher` while submitting a topology and `Scheduler` while updating the topology). In this case, `PhysicalPlan` is also getting updated my multiple components; `RuntimeManager` during kill, TMaster during submit and UpdateManager during update. Is it a concern?